### PR TITLE
[GEOS-10762] Allow enabling auto-escaping for WMS GetFeatureInfo HTML templates

### PR DESCRIPTION
**Pull request by sikeoka**
[GEOS-10762](https://osgeo-org.atlassian.net/browse/GEOS-10762)

This PR adds a WMS setting to enable/disable FreeMarker's built-in auto-escaping functionality for WMS GetFeatureInfo HTML templates.

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by sikeoka** [comment]
Any thoughts about this @aaime?

**Comment by aaime** [comment]
Will review next week, this Monday I had other engagements.

File: src/wms/src/main/java/org/geoserver/wms/WMSInfoImpl.java [hunk 1]
```diff
@@ -84,6 +84,8 @@ public class WMSInfoImpl extends ServiceInfoImpl implements WMSInfo {
 
     private Boolean defaultGroupStyleEnabled;
 
+    private boolean autoEscapeTemplateValues;
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/WMSInfoImpl.java:87]: I was about to ask "what about equals/hashcode/toString"?
But it seems none of them has been implemented in WMSInfoImpl... oooh well, carry on...

**Review by aaime — CHANGES_REQUESTED** [review]
Looks good, see comments below for minor issues.

File: doc/en/user/source/services/wms/webadmin.rst [hunk 2]
```diff
@@ -187,6 +187,15 @@ By default GetFeatureInfo results are reproject to the map coordinate reference
 
 When the flag is checked, GetFeatureInfo requests results will not be reprojected and will instead used the layer coordinate reference system.
 
+Enabling GetFeatureInfo requests results HTML auto-escaping
+-----------------------------------------------------------
+
+By default GetFeatureInfo results are printed in the HTML templates without any automatic escaping, which could result in incorrect and potentially malicious results. This behavior can be activated on a global or per virtual service basis in the **GetFeatureInfo results auto-escaping** section.
```
[the remark this answers is not in the document]
Comment (sikeoka) [reply doc/en/user/source/services/wms/webadmin.rst:193]: I'm not going to post the information here but I sent an email to the geoserver-security list a little over a month ago that demonstrated this against a GeoServer instance with default settings and no extensions and without requiring any administrator privileges.

[the remark this answers is not in the document]
Comment (aaime) [reply doc/en/user/source/services/wms/webadmin.rst:193]: Ah, I never saw the mail. Read it now.

**Review by aaime — APPROVED** [review]

**Pull request merged**